### PR TITLE
fix: allows user-agent header with header provider

### PR DESCRIPTION
**Pull request by thiagotnunes**
A bug was introduced, where if the caller tried to set a custom user agent with a header provider an exception would be thrown (for duplicate keys). Here, we merge the user agent set by the client along with the one set by the library, instead of throwing such exception.

Fixes https://github.com/googleapis/java-spanner/issues/880

### CODE REVIEW

**Comment by thiagotnunes** [comment]
cc @allenpradeep @mohanli-ml

**Comment by mohanli-ml** [comment]
LGTM

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpcTest.java [hunk 1]
```diff
@@ -502,6 +505,30 @@ public void testAdminRequestsLimitExceededRetryAlgorithm() {
...
+    final SpannerOptions options =
+        createSpannerOptions().toBuilder().setHeaderProvider(userAgentHeaderProvider).build();
+    final Spanner spanner = options.getService();
+    final DatabaseClient databaseClient =
+        spanner.getDatabaseClient(DatabaseId.of("[PROJECT]", "[INSTANCE]", "[DATABASE]"));
+    try (final ResultSet rs = databaseClient.singleUse().executeQuery(SELECT1AND2)) {
+      rs.next();
+    }
+
+    assertThat(seenHeaders.get(Key.of("user-agent", Metadata.ASCII_STRING_MARSHALLER)))
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpcTest.java:528]: nit: could we add an assertion that the default header provider is also present in the seen headers, so we prevent any silent regression from the current behavior?

**Review by olavloite — APPROVED** [review]

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpcTest.java (hunk 1 above)
Comment (thiagotnunes) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpcTest.java:528]: Good idea, I will add that.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java [hunk 2]
```diff
@@ -494,6 +489,16 @@ public GapicSpannerRpc(final SpannerOptions options) {
     }
   }
 
+  private static HeaderProvider headerProviderWithUserAgentFrom(HeaderProvider headerProvider) {
+    final Map<String, String> headersWithUserAgent = new HashMap<>(headerProvider.getHeaders());
+    final String userAgent = headersWithUserAgent.get(USER_AGENT_KEY);
```
Comment (meltsufin) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:494]: @thiagotnunes `user-agent` and `User-Agent` should be equivalent.
This needs case insensitivity.

Comment (thiagotnunes) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java:494]: Hey there, thanks for pointing this out.

This has been fixed in the following versions:

1. [3.3.3](https://github.com/googleapis/java-spanner/blob/3.3.x/CHANGELOG.md#333-2021-05-07)
1. [4.0.3](https://github.com/googleapis/java-spanner/blob/4.0.x/CHANGELOG.md#403-2021-05-03)
1. [5.2.1](https://github.com/googleapis/java-spanner/blob/5.2.x/CHANGELOG.md#521-2021-05-03)
1. [6.4.0](https://github.com/googleapis/java-spanner/blob/master/CHANGELOG.md#640-2021-04-29)

**Pull request merged**